ITISFoundation/osparc-simcore
# ✨ Is765/tags for groups  (⚠️ devops)

## What do these changes do?

🏷️ Tags are now associated to groups (instead of users) with given *read/write/delete* access rights.


### 🎉 Highlights

- ``package/postgres-database``
    - ✨ New ``models.tags_to_groups`` table that includes access rights to tags for groups
    - 🗑️ Removes ``models.tags.user_id`` column
    - ✨ New ``utils_tags`` implements repository pattern, errors and data structures for ``models.tags``
    - ♻️ cleaned up tests
- ``services/web/server``
   - ♻️ Refactors ``tag_handlers`` implementation (NOT the API!):
      - new implementation using tags repository from ``simcore_postgres_database.utils_tags``
      - improved error handling 
    - ♻️ cleaned up tests   
- ``scripts``
   -  🔨 convenience script to clone docker volumes
   - 🔨 fixes ``packages/postgres-database/scripts/erd/main.py``
   
![postgres-database-models](https://user-images.githubusercontent.com/32402063/202910027-e86365bd-fb3b-4ea6-96b4-e05e1067eb5d.png)

### ⚠️ devops
 - changes in database (see above)
 - manually tested migration against master deploy data. 
 - I recommend to do the same against the data on other deploys  (see steps below).


## Related issue/s

- ITISFoundation/osparc-issues#765: *Studies/Projects Templates/Tutorials organization* item
   - Next steps:
        - ✨ Change ``tag_handlers`` API to return associated groups and permissions (define specs w/ @odeimaiz )
        - ♻️ Improve requests validation of ``tag_handlers`` 

## How to test

```
cd packages/postgres-database
make install-dev
pytest -vv tests/test_*tag*.py
```
```
cd service/web/server
make install-dev
pytest -vv tests/**/test_*tag*.py
```

### Manual test against a deployed database
To test new migration upgrade/downgrade script:

1. clone volume of any deployed database
2. start database on cloned volume. Check ``tags`` table
3. ``sc-pg discover -u $USER -p $PASSWORD`` should detect started database
4. ``sc-pg upgrade`` should run w/o errors. 
     - Visually check ``tags`` table and new ``tags_to_groups`` table.
     - Check that ``tags.user_id`` was dropped and ``tags_to_groups`` has equivalent groups for tags
5. ``sc-pg downgrade`` 
    - Check that ``tags.user_id`` is back with correct values 

```
cd packages/postgres-database
# step 1
export POSTGRES_DATA_VOLUME=test_data_postgres_volume
make import-db-from-docker-volume host=$REMOTE_HOSTNAME host_volume=$REMOVE_VOLUME local_volume= $POSTGRES_DATA_VOLUME
 
# step 2
make up-prod

# setp 3 and 4
make migrate

# step 5
sc-pg downgrade
```


## Checklist


- [x] ``make version-*``
- [x] ``make openapi.json``
- [x] ``cd packages/postgres-database``, ``make setup-commit``, ``sc-pg review -m "my changes"``
